lib/package_utils.rb

```diff
@@ -3,10 +3,20 @@
 require 'json'
 require_relative 'const'
 
+def keep_keys(arr, keeper_keys)
+  keepers = keeper_keys.to_set
+  arr.map { |h| h.select { |k, _| keepers.include?(k) } }
+end
+
 class PackageUtils
-  def self.installed?(pkg_name)
+  def self.generate_installed_packages
     device_json = JSON.load_file(File.join(CREW_CONFIG_PATH, 'device.json'))
-    return device_json['installed_packages'].any? { |elem| elem['name'] == pkg_name }
+    return keep_keys(device_json['installed_packages'], ['name']).flat_map(&:values).to_set
```

Comment:
Given that we have to iterate over the array anyways to convert it to a set, we aren't actually getting speedups here, unfortunately:
![image](https://github.com/user-attachments/assets/c249ac11-40d3-48c7-943c-3cf6bac2c109)
```ruby
require 'json'
require 'set'
require 'benchmark'

CREW_CONFIG_PATH = '/usr/local/etc/crew'

# Method 1: Direct check for package name
def method_1(pkg_name)
  device_json = JSON.load_file(File.join(CREW_CONFIG_PATH, 'device.json'))
  device_json['installed_packages'].any? { |elem| elem['name'] == pkg_name }
end

# Method 2: Using keep_keys to filter and check
def keep_keys(arr, keeper_keys)
  keepers = keeper_keys.to_set
  arr.map { |h| h.select { |k, _| keepers.include?(k) } }
end

def method_2(pkg_name)
  device_json = JSON.load_file(File.join(CREW_CONFIG_PATH, 'device.json'))
  target = keep_keys(device_json['installed_packages'], ['name']).flat_map(&:values).to_set
  target.include?(pkg_name)
end

# Package name to search for
pkg_name = 'example_package_name'
installed_pkg_name = 'ruby'

# Benchmarking the two methods
Benchmark.bm do |x|
  x.report("Method 1, not installed package: ") { 1000.times { method_1(pkg_name) } }
  x.report("Method 2, not instapped package: ") { 1000.times { method_2(pkg_name) } }
  x.report("Method 1, installed package: ") { 1000.times { method_1(installed_pkg_name) } }
  x.report("Method 2, instapped package: ") { 1000.times { method_2(installed_pkg_name) } }
end
```

Reply:
Trying to only iterate over the array once... That means creating sets for installed packages once.

Reply:
So the idea is to have the one installed_packages set being used for multiple `installed?` checks.

Can you benchmark that?

Reply:
Let's figure this portion out another day, since this is only in `lib/fixup.rb` for now...